File: tests/test_mpi3rma_fortran/put_fence.f90
      subroutine test_mpi3rma_put_fence( base, array_size, DIM1, win, comm )

      use utilities, only: check_fompi_status

#if INCLUDE_TYPE==1
      implicit none

      include 'fompif.h'
#endif
#if INCLUDE_TYPE==2
      use fompi

      implicit none
#endif

      integer, intent(in) :: array_size
      integer, dimension(array_size), intent(inout) :: base
      integer, intent(in) :: DIM1
      integer, intent(in) :: win
      integer, intent(in) :: comm

      integer :: commsize
      integer :: rank
      integer :: mpierr
      integer :: i
      integer(kind=MPI_ADDRESS_KIND) :: start

      logical :: flag
      integer(kind=MPI_ADDRESS_KIND) :: create_flavor

      character(50) :: filename

      write (filename,*) "put_fence.f90"

      call MPI_Comm_rank( comm, rank, mpierr )
      call MPI_Comm_size( comm, commsize, mpierr )
  
      do i=1,array_size
        base(i) = commsize+i
      enddo

      call foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, create_flavor, flag, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 36, rank)
      if (flag .neqv. .true.) then
        write (*,*) "Didn't found foMPI_WIN_CREATE_FLAVOR keyval."
        STOP
      endif

      if ( create_flavor .eq. foMPI_WIN_FLAVOR_DYNAMIC ) then
        call foMPI_Get_address( base(DIM1+1), start, mpierr )
        call MPI_Sendrecv_replace( start, 1, MPI_AINT, commsize - rank - 1, 0, &
          commsize - rank - 1, 0, comm, MPI_STATUS_IGNORE, mpierr )
      else
        start = DIM1
      endif

      call foMPI_Win_fence( foMPI_MODE_NOPRECEDE, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 50, rank)
      call foMPI_Win_fence( 0, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 52, rank)

      call foMPI_Put( base(1), DIM1, MPI_INTEGER, commsize - rank - 1, start, DIM1, MPI_INTEGER, win, mpierr)
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 55, rank)

      call foMPI_Win_fence( 0, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 58, rank)

      do i=1,DIM1
        if(base(i) .ne. commsize+i) then
          write (*,*) "Error on rank ", rank, " in ", trim(filename), &
            " line ", 61, ": element ", i, " is ", base(i), &
            " instead of ", commsize+i, "."
          STOP
        endif
      enddo
      do i=1,DIM1
        if(base(DIM1+i) .ne. commsize+i) then
          write (*,*) "Error on rank ", rank, " in ", trim(filename), &
            " line ", 69, ": element ", i, " is ", base(DIM1+i), &
            " instead of ", commsize+i, "."
          STOP
        endif
      enddo
      do i=2*DIM1+1,3*DIM1
        if(base(i) .ne. commsize+i) then
          write (*,*) "Error on rank ", rank, " in ", trim(filename), &
            " line ", 78, ": element ", i, " is ", base(i), &
            " instead of ", commsize+i, "."
          STOP
        endif
      enddo

      call MPI_Barrier( comm, mpierr )
  
      if ( rank .eq. 0 ) then
        write (*,*) "No Error"
      endif

     end subroutine test_mpi3rma_put_fence 

File: tests/test_foMPI-NA/put_fence.c
#include <assert.h>

#include "fompi.h"
#include "test_mpi3rma.h"

void test_mpi3rma_put_fence( int* base, int DIM, foMPI_Win win, MPI_Comm comm ) {

  int commsize;
  int rank;
  int status;
  int i;
  MPI_Aint start;

  int flag, create_flavor;

  MPI_Comm_rank( comm, &rank );
  MPI_Comm_size( comm, &commsize );
  
  for( i=0 ; i<3*DIM ; i++) {
    base[i] = commsize+i;
  }

  status = foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, &create_flavor, &flag );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
  assert( flag != 0 );
  
  if ( create_flavor == foMPI_WIN_FLAVOR_DYNAMIC ) {
    MPI_Get_address( &base[DIM], &start);
    MPI_Sendrecv_replace( &start, 1, MPI_AINT, commsize - rank - 1, 0, commsize - rank - 1, 0, comm, MPI_STATUS_IGNORE );
  } else {
    start = DIM;
  }


  status = foMPI_Win_fence( 0, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  status = foMPI_Put( &base[0], DIM, MPI_INT, commsize - rank - 1, start, DIM, MPI_INT, win);
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  status = foMPI_Win_fence( 0, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  for( i=0 ; i<DIM ; i++) {
    if(base[i] != commsize+i) {
      fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, i, base[i], commsize+i);
      exit(EXIT_FAILURE);
    }
  }
  for( i=0 ; i<DIM ; i++) {
    if(base[DIM+i] != commsize+i) {
      fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, DIM+i, base[DIM+i], commsize+i);
      exit(EXIT_FAILURE);
    }
  }
  for( i=2*DIM ; i<3*DIM ; i++) {
    if(base[i] != commsize+i) {
      fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, i, base[i], commsize+i);
      exit(EXIT_FAILURE);
    }
  }
  if ( rank == 0 ) {
    printf("No Error\n");
  }

  return;
}
